### DISCUSSION TITLE
How to connect my PHP Container in code-server?

### DISCUSSION THREAD

**Discussion by dxkrs** [coder/code-server#6527] [Q&A]
My remote server run some containers to  service my coding work , such as nginx, mysql, php5, php7, php8. My code projects are on the server, and they mout on different php's container because of evey project maybe need the different php environment(such as PHP version, php extension), I have to isolate them.

I try to work on desktop vscode by the vscode extension--Dev Containers(ms-vscode-remote.remote-containers) a long time, it work! I can see the way is :
ssh connect my remote server in vscode's remote explorer, then it show the list of my containers on the remote server, if I chlick one of my PHP container, it will go into the container and open my project in new windows, I can get the php environment in this windows( some vscode extension such as PHP IntelliSense need it). 

I know this way is like to remote ssh in container to dev, and it's perfect for my coding.

However, sometimes I need to code without a device that can run desktop vscode, then I found this awesome project and deloy it on my remote server by docker. 

Now I can use my code-server to edit my project files, but the promble is I can't get the php environment ! The vscode extension--Dev Containers not show on the code-server extension market. I have seen some issue about "code-server ssh remote" but found the way.

So, how to connect my PHP Container in my code-server? Or other suggests?

**Comment by code-asher** [comment]
The remote SSH and containers extensions do not support code-server unfortunately, and they might never (Microsoft only allows it to run on their own desktop VS Code product at the moment).

The only way to do this with code-server would be to install code-server on the container itself then access it with a browser through a port or something like that, but that means you need a separate code-server for every container.

**Reply by dxkrs** [reply]
Microsoft not support the SSH in the web vscode, but they provides an other way to do that -- the [remote tunnels](https://code.visualstudio.com/docs/editor/vscode-web#_use-your-own-compute-with-remote-tunnels)

I try to install code-server on the container, but sometimes it failed. Some of the offical docker image which base on the old system, such as php base on alpine 3.16, they do not meet the requirements for installing code-server, such as the nodejs version or the other requirements.

Last weekend I tried to learning and use the Coder, I found that it is also achieved by installing code server in the container. However, it is hard for me to learning about the Terraform template so I had to give it up.

Finally, I want to know more the remote tunnels and want to know can it be implemented on code server?

**Reply by code-asher** [reply]
Ahh this is great information, thank you!

It looks like the remote tunnels also work through a closed-source extension, so I think we have the same problems as the other extensions.  We either have to re-implement the extension(s) with an open-source license or we have to figure out how to make the existing extensions work, which is tricky since they are closed-source and because running them in code-server is (I believe) a violation of their licenses.  In any case, someone will need to do more investigation but I believe it should be possible one way or the other.

I will add the tunnel extension to this issue: https://github.com/coder/code-server/issues/1315

**Reply by dxkrs** [reply]
Okay, I think it is really tricky beacuse of the #1315 has been discussed for more than two years. Thanks to those who did the investigation.
For now, I will try hard the method of installing code-server on the different container, and I have meet different problem to do this such as the #6535 . I will continue to explore this method.(Emmm...maybe it is the only way for now)
